### PR TITLE
Fixes #217- Added support to validate Begin/End date for Excel Import.

### LINKED ISSUE #217: DAMS Excel Import allows invalid Begin/End date - object is not visible
### Descriptive summary

Currently, pre-ingest and ingest validation does not validate Begin date and End date values.  For example, 2018-02-31 is invalid because February does not have 31 days.
In addition to the current ISO date format (YYYY-MM-DD) validation, date values should also be validated.

### Rationale

This is a problem because an object containing an invalid Begin or End date value is not visible in the DAMS public or curator view, meaning the object cannot be found and inspected.  It might be that SOLR cannot index the Object.  When a large collection is ingested and the final object count is short, it is very difficult (and very time consuming) to determine which objects are missing and where the error is so they can be edited.

### Expected behavior
pre-ingest and ingest fail and report is sent: "Invalid begin date 2018-02-31 in record 10001. Must be a valid calendar date."
Rules:
MM must be at least 01 and no greater than 12
DD must be at least 01 and no greater than the maximum days for the month expressed in MM (must account for leap years)

January <= 31 days
February <= 28 days, 29 in leap years
March <= 31 days
April <= 30 days
May <= 31 days
June <= 30 days
July <= 31 days
August <= 31 days
September <= 30 days
October <= 31 days
November <= 30 days
December <= 31 days

### Actual behavior
pre-ingest and ingest succeed and object is not found in the DAMS (public, curator, data view, rdf view)

### Examples

Invalid Begin date of 2018-02-31. Not visible in curator mode
http://librarytest.ucsd.edu/dc/object/bd34316619
http://librarytest.ucsd.edu/dc/object/bd34316619/data
http://librarytest.ucsd.edu/dc/object/bd34316619/rdf

Valid Begin date of 2018-02-28. Visible in curator mode
http://librarytest.ucsd.edu/dc/object/bd0155180z
http://librarytest.ucsd.edu/dc/object/bd0155180z/data
http://librarytest.ucsd.edu/dc/object/bd0155180z/rdf

### PR DESCRIPTION
**Pull request by lsitu**
Fixes #217
Validate Begin/End date for Excel Import.

@ucsdlib/developers - please review

### CODE REVIEW

File: src/edu/ucsd/library/xdre/tab/ExcelSource.java [hunk 1]
```diff
@@ -40,6 +42,10 @@
 	public static final String[] IGNORED_FIELDS_FOR_OBJECTS = {"CLR image file name", "Brief description", "subject type"};
 	public static final String[] IGNORED_FIELDS_FOR_COLLECTIONS = {"Level","Title","Subtitle","Part name","Part number","Translation","Variant","File name","File use","File name 2","File use 2", "subject type"};
 
+	private static final String DATETIME_FORMAT = "yyyy-MM-dd";
```
Comment (mcritchlow) [inline src/edu/ucsd/library/xdre/tab/ExcelSource.java]: seems like too large an indent here?

File: src/edu/ucsd/library/xdre/tab/ExcelSource.java [hunk 2]
```diff
@@ -248,23 +254,49 @@ else if ( cache != null && cache.size() > 0)
...
-			    	                else
-			    	                	invalids.put(originalHeader, existing + " " + DELIMITER + " " + val);
+	                        if (originalHeader.equalsIgnoreCase(BEGIN_DATE) || originalHeader.equalsIgnoreCase(END_DATE))
+	                        {
+	                            // validate date-time format
+	                            boolean validDate = false;
+	                            StringBuilder messageBuilder = new StringBuilder();
+	                            for (String dateFormatter : validValue)
+	                            {
+		                            try {
```
Comment (mcritchlow) [inline src/edu/ucsd/library/xdre/tab/ExcelSource.java]: try block seems indented more than other code?

File: src/edu/ucsd/library/xdre/tab/ExcelSource.java [hunk 3]
```diff
@@ -248,23 +254,49 @@ else if ( cache != null && cache.size() > 0)
 	                    List<String> validValue = CONTROL_VALUES.get(header);
 	                    if (validValue != null && validValue.size() > 0)
```
Comment (mcritchlow) [inline src/edu/ucsd/library/xdre/tab/ExcelSource.java]: Possible to just return here if `validValue` is `null`? Might let you collapse one conditional path at least.

File: src/edu/ucsd/library/xdre/tab/ExcelSource.java [hunk 4]
```diff
@@ -248,23 +254,49 @@ else if ( cache != null && cache.size() > 0)
...
+	                            }
+
+	                            if (!validDate)
+	                            {
+	                                invalids.put(originalHeader, messageBuilder.toString());
+	                            }
+	                        }
+	                        else
+	                        {
+		                    	// validate cell with multiple values
```
Comment (mcritchlow) [inline src/edu/ucsd/library/xdre/tab/ExcelSource.java]: indention/spacing of this section seems inconsistent too. It seems like there is 2, 4, and 8 space indention. Particularly the if/else if checking Language below.

**Comment by lsitu** [comment]
@mcritchlow It's done. Thanks.

**Pull request merged**